### ISSUE TITLE
[REVIEW]: Geometric nested sampling

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1809]
**Submitting author:** @SuperKam91 (<a href="http://orcid.org/0000-0002-2089-3964">Kamran Javid</a>)
**Repository:** <a href="https://github.com/SuperKam91/gns" target ="_blank">https://github.com/SuperKam91/gns</a>
**Version:** v1.01
**Editor:** @mbobra
**Reviewer:** @ziotom78, @zhampel
**Archive:** <a href="https://doi.org/10.5281/zenodo.3569956" target="_blank">10.5281/zenodo.3569956</a>

## Status

[![status](https://joss.theoj.org/papers/34c9c9b4b15a502199f0b4c68ec7e48b/status.svg)](https://joss.theoj.org/papers/34c9c9b4b15a502199f0b4c68ec7e48b)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/34c9c9b4b15a502199f0b4c68ec7e48b"><img src="https://joss.theoj.org/papers/34c9c9b4b15a502199f0b4c68ec7e48b/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/34c9c9b4b15a502199f0b4c68ec7e48b/status.svg)](https://joss.theoj.org/papers/34c9c9b4b15a502199f0b4c68ec7e48b)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/SuperKam91/gns" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@ziotom78 & @zhampel, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @mbobra know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/SuperKam91/gns">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@SuperKam91) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/SuperKam91/gns">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@SuperKam91) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @ziotom78, @zhampel it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01809/joss.01809/10.21105.joss.01809.pdf)

**Comment by mbobra** [comment]
👋@ziotom78 @zhampel Thank you again for volunteering to review. The review instructions are listed above. Please let me know if you have any questions!

**Comment by ziotom78** [comment]
I am still completing the review of the code, but I started reading the paper as well. One thing I note is Fig. 2, which has two "X" marks that are explained neither in the caption nor in the text. The figure is obviously Fig. 3 in the [arXiv paper 1905.09110](https://arxiv.org/pdf/1905.09110.pdf); however, in the arXiv paper Fig. 3 makes sense because it is better explained and placed alongside Fig. 2. I believe that you should include the same figure and the same level of detail in this paper as well.

If the paper length is an issue, you could scale down Fig. 2 in the JOSS paper a bit: the information content is very small (a sphere, three points, and three labels), but the space it takes is really huge.

**Comment by SuperKam91** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01809/joss.01809/10.21105.joss.01809.pdf)

**Comment by SuperKam91** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01809/joss.01809/10.21105.joss.01809.pdf)

**Comment by SuperKam91** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01809/joss.01809/10.21105.joss.01809.pdf)

**Comment by zhampel** [comment]
@SuperKam91 As I begin my review, can you please include some guidelines for contributing or reporting issues to your codebase? Here's an [example](https://jrbourbeau.github.io/pyunfold/contributing.html).

**Comment by SuperKam91** [comment]
> I am still completing the review of the code, but I started reading the paper as well. One thing I note is Fig. 2, which has two "X" marks that are explained neither in the caption nor in the text. The figure is obviously Fig. 3 in the [arXiv paper 1905.09110](https://arxiv.org/pdf/1905.09110.pdf); however, in the arXiv paper Fig. 3 makes sense because it is better explained and placed alongside Fig. 2. I believe that you should include the same figure and the same level of detail in this paper as well.
> 
> If the paper length is an issue, you could scale down Fig. 2 in the JOSS paper a bit: the information content is very small (a sphere, three points, and three labels), but the space it takes is really huge.

Hi @ziotom78 thank you for your comments on the paper. While I gave some information on the coordinate labels in the caption of Figure 2 (n.b. captions should be visible in the `.pdf`s generated in this thread, but probably not in the `.md` file itself), I have added a couple of sentences in the main text elaborating more on them. With regards to the figure sizes, I have tried to reduce them using different syntaxes to the one suggested in the JOSS `paper.md` example. These worked in the `.md` files, but not in the `.pdf` generated here, and so I have had to revert back to the original sizing. It would be very useful if the JOSS template provided an example of how to scale images which is compatible with the `pdf` generation used here.

**Comment by SuperKam91** [comment]
Hi @zhampel , I have added a section explaining how to contribute to the project/ report issues at the bottom of the `README` of the repo.

**Comment by ziotom78** [comment]
As issue [#3](https://github.com/SuperKam91/gns/issues/3) is now closed, I think that the paper and the package are ok, but let's wait for @zhampel's approval too. The functionality provided by the package looks good, nice work!

However, there are a few suggestions I would like to give to @SuperKam91 :

- The API documentation is ok, but it is surely not «good» [according to JOSS' guidelines](https://joss.readthedocs.io/en/latest/review_criteria.html#api-documentation):

    - It is just a list of functions provided by the package, with no example for any of them. Compare this with the documentation of the Python's standard library (e.g., the [datetime module](https://docs.python.org/3/library/datetime.html)): short code snippet explain how each function is meant to be used.

    - As the list of functions provided by the package is very long, some text introducing each module (e.g., `gns.keeton_calculations`) would surely have helped interested readers. Again, Python's documentation is a good example to be used as reference.

- The README is overly long, and it provides long textual descriptions of how to use the library, when short Python snippets would have provided the same information in a much clearer form.

- The review process was quite long because there were trivial errors in the source code that should have been spotted by the author.

In the future, please read carefully [Author and Reviewer Guides](https://joss.readthedocs.io/en/latest/index.html) before submitting a paper to JOSS, as this will ensure a much faster and smoother review process.

**Comment by mbobra** [comment]
👋 @zhampel Let me know if you have any questions or need any help completing your review!

**Comment by zhampel** [comment]
@SuperKam91 
I agree with @ziotom78 as well. Your submission's functionality is as advertised, but the documentation is lacking in detail. For example, while not explicitly required, function input types and descriptors provide users context for implementation. This is also reflected in the difficulty in scanning the example code in the README, which I would not characterize as a usage guide.

With respect to testing, it doesn't appear that the scripts in the `test` directory strictly qualify as unit tests, specifically the lack of assertion type tests. This would also be facilitated or at least clarified by care detailing of input argument types and type-checks as appropriate. This is also evident in abundant commented code ending with `#debug` or similar. For example on line 120 of `nested_run.py`:

```       #elif setupDict['sampler'] == 'MH' or setupDict['sampler'] == 'MH geom': #temporary whilst testing geometric ns```

If such lines are intended for debugging or testing, then the preference would be to add in some testing capability or flags to ensure such tests are performed with requisite feedback.

Furthermore, I think the paper and docs still have some errors. Examples:
- on page 3: "For example, parameters which... the trial distribution..."
  The sentence is not complete as 'For example' and 'the trial distribution' do not relate properly.
  You can simply state 'periodic properties' instead of 'circular or periodic'
- In the 4th paragraph of the online docs: "Note for sampling spherical sampling, two physical parameters..."
- You also use quotes frequently when they are not necessary.

I recommend that the written components of the submission be re-reviewed and corrected, as well as addressing the lack of clear comments as well as copious commented code within the main codebase. I also highly recommend use of a style guide for Python code such as [pep8](https://www.python.org/dev/peps/pep-0008/).

**Comment by SuperKam91** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01809/joss.01809/10.21105.joss.01809.pdf)

**Comment by SuperKam91** [comment]
> @SuperKam91
> I agree with @ziotom78 as well. Your submission's functionality is as advertised, but the documentation is lacking in detail. For example, while not explicitly required, function input types and descriptors provide users context for implementation. 


Hi @zhampel thank you for your comments. I will address them sequentially:

> This is also reflected in the difficulty in scanning the example code in the README, which I would not characterize as a usage guide.

I'm not quite sure I follow unfortunately, are you saying the example needs more comments to explain what each line of code does? Anyway, I have added more documentation to the `toy_models.py` sub-package which is called from the example.

> With respect to testing, it doesn't appear that the scripts in the `test` directory strictly qualify as unit tests, specifically the lack of assertion type tests. 

While I didn't see anything in the JOSS guidelines specifying that unit tests were strictly necessary, I appreciate your point, and also acknowledge the fact that a lot of JOSS repos do include them, and therefore I have included some in `tests.py`. For a number of toy models, these essentially check that the statistics associated with the nested sampling runs are calculated correctly.

This would also be facilitated or at least clarified by care detailing of input argument types and type-checks as appropriate. This is also evident in abundant commented code ending with `#debug` or similar. For example on line 120 of `nested_run.py`:
> 
> ` #elif setupDict['sampler'] == 'MH' or setupDict['sampler'] == 'MH geom': #temporary whilst testing geometric ns`
> 
> If such lines are intended for debugging or testing, then the preference would be to add in some testing capability or flags to ensure such tests are performed with requisite feedback.

I have taken care of such redundant/commented out code.

> Furthermore, I think the paper and docs still have some errors. Examples:
> 
>     * on page 3: "For example, parameters which... the trial distribution..."
>       The sentence is not complete as 'For example' and 'the trial distribution' do not relate properly.
>       You can simply state 'periodic properties' instead of 'circular or periodic'
> 
>     * In the 4th paragraph of the online docs: "Note for sampling spherical sampling, two physical parameters..."

I have addressed these two typos, thank you for pointing them out.

> I recommend that the written components of the submission be re-reviewed and corrected, as well as addressing the lack of clear comments as well as copious commented code within the main codebase. I also highly recommend use of a style guide for Python code such as [pep8](https://www.python.org/dev/peps/pep-0008/).

I appreciate the code API may not be very good in certain places, but as I said to @ziotom78 in a closed issue, I have ensured that the API is far superior for all the sub-packages/functions which an end-user is typically going to interact with, i.e. the sub-packages:
`nested_run.py`
`plotting.py`
`output.py`
`theoretical_funcs.py`

While I would love to re-write the package following pep8, I think it is unlikely I will get time to.

On another note, I added an extra sentence to the paper explaining where the relative performances of GNS and another popular nested sampling algorithm, MultiNest, can be found (at the end of the section Bayesian inference and nested sampling).

**Comment by zhampel** [comment]
@SuperKam91 I pulled your most recent changes, but I would not agree with you statement that
> I have taken care of such redundant/commented out code.

Looking again in `nested_run.py`, I still see a significant number of commented full lines with qualifications like 'temporary' and 'debug'.

It would be worth you intended users' time to clarify your code as much as possible.

**Comment by ziotom78** [comment]
> While I would love to re-write the package following pep8, I think it is unlikely I will get time to.

No need to rewrite your package, there are automatic tools like [autopep8](https://github.com/hhatto/autopep8), [yapf](https://github.com/google/yapf) and [black](https://github.com/psf/black) that can do the job for you.

**Comment by SuperKam91** [comment]
Thank you for you reply @zhampel, that seems very strange. Did you pull from PyPi? I haven't added any of the the review-changes to that version of the package yet, as I figured I would do it at the end of the review process (to avoid having to keep releasing new versions). Please use the version in the github repo for now. 
If not I can only think that you looked between me posting the above comment and me uploaded the changes, but they are definitely present in the repo now.

@ziotom78 thank you for the heads up, those are truly great packages! Consider my code now *almost* fully PEP8 compliant- the only places where it isn't is for lines which include comments, which the packages cannot linebreak. I ran the pep8 test and got the following output:
1106    E501 line too long (96 > 79 characters)
Hopefully you will spare me the pain of line breaking my 1106 lines of comments!

**Comment by mbobra** [comment]
@SuperKam91 There are tools like [autopep8](https://pypi.org/project/autopep8/) that will automatically format your code for you so you don't have to do anything by hand :)

**Comment by SuperKam91** [comment]
> @SuperKam91 There are tools like [autopep8](https://pypi.org/project/autopep8/) that will automatically format your code for you so you don't have to do anything by hand :)

Hi @mbobra as I mentioned above, I have already used tools including ‘autopep8’, ‘black’ and ’yapf’ to arrive at the current state. It is fully pep8 compliant apart from my comment lines being over the 80 characters per line limit, and also my doc strings being over the limit. Understandably, the tools cannot correctly break all the commented lines, especially lines which have comments proceeding code.  But since these are just comments I don’t think it should do much harm.

**Comment by SuperKam91** [comment]
Hi @mbobra @zhampel any updates on this? Please see my above posts which I hope address your comments to a satisfactory level.

**Comment by mbobra** [comment]
@zhampel Does this submission pass your review? If so, can you please approve this for publication (tick all the checkboxes) or let @SuperKam91 know what issues still remain?

**Comment by mbobra** [comment]
@zhampel Please let me know if you need any help assessing whether this submission passes review! I am happy to help.

**Comment by zhampel** [comment]
@SuperKam91 

I apologize for my delay in coming back to your submission.

Thanks for addressing most of the issues. However, I still feel like leaving extremely long comments in **code** sets a negative precedent for JOSS accepted submissions. You wrote the following:
> But since these are just comments I don’t think it should do much harm.

yet, I see this massive single line (327) in tests/test.py:

`setupDict = {'verbose': True, 'trapezoidalFlag': False, 'ZLiveType':'average X', terminationType':'evidence', 'terminationFactor': 0.01, 'sampler': None, 'outputFile': None, 'space':'linear', 'paramGeomList':None}
`

among others. While I agree that long docstrings do not do much harm, I argue that if you want your novel code to be useful to others, with which a publication through JOSS facilitates and which provides authors official credit, then it's best to abide by coding conventions. They are there for a reason, to be helpful for others. And that's why there are automated tools to implement those conventions.


@mbobra 
I have reluctantly checked all the boxes and added an issue ticket for the long lines (a more appropriate documentation for this discussion). The state of this codebase is not up to my standards but it is within the guidelines for publication. I would like to stress that reviewers volunteer their time for submissions, and if authors do not take reviewers' time and experience into consideration, then I think it also sets a poor precedent for acceptance. This can affect not just the reviewing process, but more importantly the end-users in the open-source community which we serve.

**Comment by mbobra** [comment]
@zhampel Thank you for your time and your well-considered review. I totally and completely understand your comments. We at JOSS are trying to strike some kind of reasonable balance between incentivizing people to write open source software (with a reasonable amount of time and effort and also) and adhering to the most ideal coding standards. I personally agree with you that long comments don't adhere to coding conventions, but I'm also bound (like you) to follow JOSS' guidelines. I hope you feel positive about your experience reviewing for JOSS. And if you feel like the JOSS guidelines should be modified, [please open an issue](https://github.com/openjournals/joss/issues/new)! Thank you sincerely for your time and effort ☀️

**Comment by mbobra** [comment]
@SuperKam91 We're almost there! Can you please [archive your release on Zenodo](https://guides.github.com/activities/citable-code/) to obtain a DOI and then put that in your README.md file? After that I think we're done 🎉

**Comment by SuperKam91** [comment]
> @SuperKam91
> 
> I apologize for my delay in coming back to your submission.
> 
> Thanks for addressing most of the issues. However, I still feel like leaving extremely long comments in **code** sets a negative precedent for JOSS accepted submissions. You wrote the following:
> 
> > But since these are just comments I don’t think it should do much harm.
> 
> yet, I see this massive single line (327) in tests/test.py:
> 
> `setupDict = {'verbose': True, 'trapezoidalFlag': False, 'ZLiveType':'average X', terminationType':'evidence', 'terminationFactor': 0.01, 'sampler': None, 'outputFile': None, 'space':'linear', 'paramGeomList':None} `
> 
> among others. While I agree that long docstrings do not do much harm, I argue that if you want your novel code to be useful to others, with which a publication through JOSS facilitates and which provides authors official credit, then it's best to abide by coding conventions. They are there for a reason, to be helpful for others. And that's why there are automated tools to implement those conventions.
> 
> @mbobra
> I have reluctantly checked all the boxes and added an issue ticket for the long lines (a more appropriate documentation for this discussion). The state of this codebase is not up to my standards but it is within the guidelines for publication. I would like to stress that reviewers volunteer their time for submissions, and if authors do not take reviewers' time and experience into consideration, then I think it also sets a poor precedent for acceptance. This can affect not just the reviewing process, but more importantly the end-users in the open-source community which we serve.

Hi @zhampel I will ensure down the road that the code is full pep8 compliant (including comment/docstring lines), even if I do this process gradually, rather than all at once. 

Thanks again for taking the time to do the review. Also thank you to @ziotom78 for his review, and to @mbobra for overseeing everything.

**Comment by mbobra** [comment]
@SuperKam91 I appreciate your effort; I think pep8 compliance would make the code much more legible. Could you please [archive your release on Zenodo](https://guides.github.com/activities/citable-code/) to obtain a DOI and then put that in your README.md file? Please make sure the title and author list of the Zenodo entry match the title and author list of the JOSS paper. Thank you!

**Comment by SuperKam91** [comment]
> @SuperKam91 I appreciate your effort; I think pep8 compliance would make the code much more legible. Could you please [archive your release on Zenodo](https://guides.github.com/activities/citable-code/) to obtain a DOI and then put that in your README.md file? Please make sure the title and author list of the Zenodo entry match the title and author list of the JOSS paper. Thank you!

Done!

**Comment by mbobra** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01809/joss.01809/10.21105.joss.01809.pdf)

**Comment by mbobra** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by whedon** [comment]
```Reference check summary:

OK DOIs

- None

MISSING DOIs

- https://doi.org/10.1080/00031305.1995.10476177 may be missing for title: Understanding the metropolis-hastings algorithm
- https://doi.org/10.1093/mnras/stz826 may be missing for title: Sunyaev–Zel’dovich profile fitting with joint AMI-Planck analysis
- https://doi.org/10.1093/mnras/stz2341 may be missing for title: Physical modelling of galaxy cluster Sunyaev–Zel’dovich data using Einasto dark matter profiles
- https://doi.org/10.1111/j.1365-2966.2009.15247.x may be missing for title: Bayesian modelling of clusters of galaxies from multifrequency-pointed Sunyaev–Zel’dovich observations
- https://doi.org/10.1111/j.1365-2966.2009.14548.x may be missing for title: MultiNest: an efficient and robust Bayesian inference tool for cosmology and particle physics
- https://doi.org/10.1214/06-ba127 may be missing for title: Nested sampling for general Bayesian computation
- https://doi.org/10.1093/mnras/stv1911 may be missing for title: POLYCHORD: next-generation nested sampling
- https://doi.org/10.1093/mnrasl/slv047 may be missing for title: PolyChord: nested sampling for cosmology
- https://doi.org/10.1111/j.1365-2966.2007.12353.x may be missing for title: Multimodal nested sampling: an efficient and robust alternative to Markov Chain Monte Carlo methods for astronomical data analyses

INVALID DOIs

- None
```

**Comment by mbobra** [comment]
@SuperKam91 Could you please include DOIs for all your references? So for example, the first entry in paper.bib would look like this: 

```
@article{chib1995understanding,
  title={Understanding the metropolis-hastings algorithm},
  author={Chib, Siddhartha and Greenberg, Edward},
  journal={The american statistician},
  volume={49},
  number={4},
  pages={327--335},
  year={1995},
  publisher={Taylor \& Francis Group},
  doi={10.1080/00031305.1995.10476177}
}
```

Thank you 😄

**Comment by mbobra** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references...
```

**Comment by SuperKam91** [comment]
Hi @mbobra apologies for the delay in including the reference DOIs. I am basically waiting on my university to provide me with a DOI for one of the cited papers, but they are taking an age to do so. As soon as they provide it I will update and the references should be good to go!

**Comment by mbobra** [comment]
@SuperKam91 Do you want to go ahead and publish the paper without the DOI for that one citation? That's fine with me -- we don't need to hold up your publication just for that singular DOI.

**Comment by SuperKam91** [comment]
@whedon check references

**Comment by SuperKam91** [comment]
Hi @mbobra in the end I've lost patience with regard to my university providing me the DOI of the reference, so I just created one for it myself using Zenodo. Hopefully this is sufficient, in which case we can proceed.

**Comment by SuperKam91** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #1809 with the following error: 

 Error reading bibliography ./paper.bib (line 10, column 3):
unexpected "d"
expecting space, ",", white space or "}"
Error running filter pandoc-citeproc:
Filter returned error status 1
Looks like we failed to compile the PDF

**Comment by SuperKam91** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1080/00031305.1995.10476177 is OK
- 10.1093/mnras/sty3115 is OK
- 10.17863/CAM.38865 is OK
- 10.1093/mnras/stz826 is OK
- 10.5281/zenodo.3627542 is OK
- 10.17863/CAM.40616 is OK
- 10.1093/mnras/stz2341 is OK
- 10.1111/j.1365-2966.2009.15247.x is OK
- 10.1111/j.1365-2966.2009.14548.x is OK
- 10.1214/06-BA127 is OK
- 10.1093/mnras/stv1911 is OK
- 10.1093/mnrasl/slv047 is OK
- 10.1111/j.1365-2966.2007.12353.x is OK
- 10.1093/mnras/sty074 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by SuperKam91** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01809/joss.01809/10.21105.joss.01809.pdf)

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1080/00031305.1995.10476177 is OK
- 10.1093/mnras/sty3115 is OK
- 10.17863/CAM.38865 is OK
- 10.1093/mnras/stz826 is OK
- 10.5281/zenodo.3627542 is OK
- 10.17863/CAM.40616 is OK
- 10.1093/mnras/stz2341 is OK
- 10.1111/j.1365-2966.2009.15247.x is OK
- 10.1111/j.1365-2966.2009.14548.x is OK
- 10.1214/06-BA127 is OK
- 10.1093/mnras/stv1911 is OK
- 10.1093/mnrasl/slv047 is OK
- 10.1111/j.1365-2966.2007.12353.x is OK
- 10.1093/mnras/sty074 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1080/00031305.1995.10476177 is OK
- 10.1093/mnras/sty3115 is OK
- 10.17863/CAM.38865 is OK
- 10.1093/mnras/stz826 is OK
- 10.5281/zenodo.3627542 is OK
- 10.17863/CAM.40616 is OK
- 10.1093/mnras/stz2341 is OK
- 10.1111/j.1365-2966.2009.15247.x is OK
- 10.1111/j.1365-2966.2009.14548.x is OK
- 10.1214/06-BA127 is OK
- 10.1093/mnras/stv1911 is OK
- 10.1093/mnrasl/slv047 is OK
- 10.1111/j.1365-2966.2007.12353.x is OK
- 10.1093/mnras/sty074 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by mbobra** [comment]
@whedon set 10.5281/zenodo.3569956 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3569956" target="_blank">10.5281/zenodo.3569956</a> is the archive.

**Comment by mbobra** [comment]
@whedon set v1.01 as version

**Comment by whedon** [comment]
OK. v1.01 is the version.

**Comment by mbobra** [comment]
@openjournals/joss-eics the paper is ready for final processing

**Comment by Kevin-Mattheus-Moerman** [comment]
@SuperKam91 I am the AEiC helping to process this submission. Below are some remaining minor points to work on in relation to your paper. 

- [ ] The first sentence is quite difficult to read. Can you consider rephrasing, splitting into multiple sentences, or adding more comma's? 
- [ ] Please change `...present here is a based on...` to `...present here is based on...`
- [ ] Fix typo in `intutive` which should be `intuitive`
- [ ] Fix typo in `equivlance` `equivalance`
- [ ] Consider changing `Figures 6 - 9 of the paper.... ` to `Figures 6 - 9 of that paper.... ` which to me makes it a bit clearer that you are referring to the before mentioned study not the current paper. 
- [ ] Please add comma in `To prevent this one must sample a large number...` i.e. `To prevent this, one must sample a large number...`
- [ ] Should `This solution can be problematic if...` be rephrased to `Obtaining this solution can be problematic if...`?
- [ ] Can you replace `...is ‘wrapped’ around...` with `...is "wrapped" around...`, so double quotation symbols are used consistently throughout the text? 
- [ ] Please spell out United Kingdom in your affiliations. 
- [ ] This paper is about to be processed for acceptance in JOSS. I recommend you also proofread the text thoroughly one last time.

**Comment by SuperKam91** [comment]
Hi @Kevin-Mattheus-Moerman, I can confirm I have addressed the changes proposed above. Furthermore I have given the paper a final proofread.

**Comment by SuperKam91** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #1809 with the following error: 

 /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:377:in `parse': (tmp/1809/paper.md): did not find expected key while parsing a block mapping at line 2 column 1 (Psych::SyntaxError)
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:377:in `parse_stream'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:325:in `parse'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:252:in `load'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:473:in `block in load_file'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:472:in `open'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:472:in `load_file'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:125:in `load_yaml'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:85:in `initialize'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `new'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `set_paper'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:55:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:116:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by SuperKam91** [comment]
@whedon check references

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
PDF failed to compile for issue #1809 with the following error: 

 /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:377:in `parse': (tmp/1809/paper.md): did not find expected key while parsing a block mapping at line 2 column 1 (Psych::SyntaxError)
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:377:in `parse_stream'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:325:in `parse'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:252:in `load'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:473:in `block in load_file'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:472:in `open'
	from /app/vendor/ruby-2.4.4/lib/ruby/2.4.0/psych.rb:472:in `load_file'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:125:in `load_yaml'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon.rb:85:in `initialize'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `new'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/lib/whedon/processor.rb:36:in `set_paper'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:55:in `prepare'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
	from /app/vendor/bundle/ruby/2.4.0/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
	from /app/vendor/bundle/ruby/2.4.0/bundler/gems/whedon-9847f98e9ec6/bin/whedon:116:in `<top (required)>'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `load'
	from /app/vendor/bundle/ruby/2.4.0/bin/whedon:23:in `<main>'

**Comment by Kevin-Mattheus-Moerman** [comment]
@SuperKam91 did you have the ` symbols in the title field for the YAML content at the top of your paper there before? 
```yaml
title: 'Geometric nested sampling: sampling from distributions defined on non-trivial geometries'
```
Perhaps change it to the following and try again:
```yaml
title: Geometric nested sampling: sampling from distributions defined on non-trivial geometries
```

**Comment by arfon** [comment]
@SuperKam91 - this PR will fix your paper compilation: https://github.com/SuperKam91/gns/pull/6

**Comment by SuperKam91** [comment]
> @SuperKam91 - this PR will fix your paper compilation: [SuperKam91/gns#6](https://github.com/SuperKam91/gns/pull/6)

So sorry about that @arfon, don't know how that occurred!

**Comment by SuperKam91** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01809/joss.01809/10.21105.joss.01809.pdf)

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1080/00031305.1995.10476177 is OK
- 10.1093/mnras/sty3115 is OK
- 10.17863/CAM.38865 is OK
- 10.1093/mnras/stz826 is OK
- 10.5281/zenodo.3627542 is OK
- 10.17863/CAM.40616 is OK
- 10.1093/mnras/stz2341 is OK
- 10.1111/j.1365-2966.2009.15247.x is OK
- 10.1111/j.1365-2966.2009.14548.x is OK
- 10.1214/06-BA127 is OK
- 10.1093/mnras/stv1911 is OK
- 10.1093/mnrasl/slv047 is OK
- 10.1111/j.1365-2966.2007.12353.x is OK
- 10.1093/mnras/sty074 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@SuperKam91 thanks for making those changes. Looks like we are all set now.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1080/00031305.1995.10476177 is OK
- 10.1093/mnras/sty3115 is OK
- 10.17863/CAM.38865 is OK
- 10.1093/mnras/stz826 is OK
- 10.5281/zenodo.3627542 is OK
- 10.17863/CAM.40616 is OK
- 10.1093/mnras/stz2341 is OK
- 10.1111/j.1365-2966.2009.15247.x is OK
- 10.1111/j.1365-2966.2009.14548.x is OK
- 10.1214/06-BA127 is OK
- 10.1093/mnras/stv1911 is OK
- 10.1093/mnrasl/slv047 is OK
- 10.1111/j.1365-2966.2007.12353.x is OK
- 10.1093/mnras/sty074 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1261

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1261, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1223503513793302528) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1262
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01809](https://doi.org/10.21105/joss.01809)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.01809/status.svg)](https://doi.org/10.21105/joss.01809)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01809">
  <img src="https://joss.theoj.org/papers/10.21105/joss.01809/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.01809/status.svg
   :target: https://doi.org/10.21105/joss.01809
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.01809/status.svg)](https://doi.org/10.21105/joss.01809)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by SuperKam91** [comment]
Many thanks all for your help, I have made a small donation to JOSS in the hope that you all can keep up the good work in the future. 

I have noticed a small typo in the references of the paper. Is it possible to update the publication now I have ratified it? 

If not absolutely no worries, I will just include it in the arXiv version instead.

**Comment by SuperKam91** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1080/00031305.1995.10476177 is OK
- 10.1093/mnras/sty3115 is OK
- 10.17863/CAM.38865 is OK
- 10.1093/mnras/stz826 is OK
- 10.5281/zenodo.3627542 is OK
- 10.17863/CAM.40616 is OK
- 10.1093/mnras/stz2341 is OK
- 10.1111/j.1365-2966.2009.15247.x is OK
- 10.1111/j.1365-2966.2009.14548.x is OK
- 10.1214/06-BA127 is OK
- 10.1093/mnras/stv1911 is OK
- 10.1093/mnrasl/slv047 is OK
- 10.1111/j.1365-2966.2007.12353.x is OK
- 10.1093/mnras/sty074 is OK

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1093/mnras/stz2341 is INVALID because of 'https://doi.org/' prefix
```

**Comment by SuperKam91** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1080/00031305.1995.10476177 is OK
- 10.1093/mnras/sty3115 is OK
- 10.17863/CAM.38865 is OK
- 10.1093/mnras/stz826 is OK
- 10.5281/zenodo.3627542 is OK
- 10.17863/CAM.40616 is OK
- 10.1093/mnras/stz2341 is OK
- 10.1111/j.1365-2966.2009.15247.x is OK
- 10.1111/j.1365-2966.2009.14548.x is OK
- 10.1214/06-BA127 is OK
- 10.1093/mnras/stv1911 is OK
- 10.1093/mnrasl/slv047 is OK
- 10.1111/j.1365-2966.2007.12353.x is OK
- 10.1093/mnras/sty074 is OK

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1093/mnras/stz2341 is INVALID because of 'https://doi.org/' prefix
```

**Comment by arfon** [comment]
> Many thanks all for your help, I have made a small donation to JOSS in the hope that you all can keep up the good work in the future.

❤️ Many thanks!

> I have noticed a small typo in the references of the paper. Is it possible to update the publication now I have ratified it?

Yes, please go ahead and make the fix in the `master` branch of your repository and mention me here (@arfon) when you've done this.

**Comment by SuperKam91** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1080/00031305.1995.10476177 is OK
- 10.1093/mnras/sty3115 is OK
- 10.17863/CAM.38865 is OK
- 10.1093/mnras/stz826 is OK
- 10.5281/zenodo.3627542 is OK
- 10.17863/CAM.40616 is OK
- 10.1093/mnras/stz2341 is OK
- 10.1111/j.1365-2966.2009.15247.x is OK
- 10.1111/j.1365-2966.2009.14548.x is OK
- 10.1214/06-BA127 is OK
- 10.1093/mnras/stv1911 is OK
- 10.1093/mnrasl/slv047 is OK
- 10.1111/j.1365-2966.2007.12353.x is OK
- 10.1093/mnras/sty074 is OK

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1093/mnras/stz2341 is INVALID because of 'https://doi.org/' prefix
```

**Comment by SuperKam91** [comment]
> > Many thanks all for your help, I have made a small donation to JOSS in the hope that you all can keep up the good work in the future.
> 
>  Many thanks!
> 
> > I have noticed a small typo in the references of the paper. Is it possible to update the publication now I have ratified it?
> 
> Yes, please go ahead and make the fix in the `master` branch of your repository and mention me here (@arfon) when you've done this.

Hi @arfon that's great, thank you so much. I have made the changes to `master` branch so we should be good to go. However, when "checking references" above it seems to have cached an old commit (perhaps I didn't wait long enough but I don't think this is usually an issue) which doesn't include the fix, so let me know if you have any problems with this.

**Comment by arfon** [comment]
OK, this should be fixed now. The PDF might take a few hours to show up as fixed online due to caching.

Also, I think you're right, there's some kind of caching issue with Whedon too that I need to look into.

**Comment by kthyng** [comment]
@arfon I've been having troubles with this too, like in #1656. I'll check it soon and it's probably now worked out but it gets confusing!
